wgpu-native/src/track/buffer.rs

```diff
@@ -0,0 +1,125 @@
+use crate::{
+    conv,
+    resource::BufferUsage,
+    BufferId,
+};
+use super::{PendingTransition, ResourceState, Stitch, Unit};
+use std::ops::Range;
+
+//TODO: store `hal::buffer::State` here to avoid extra conversions
+pub type BufferState = Unit<BufferUsage>;
+
+impl PendingTransition<BufferState> {
+    /// Produce the gfx-hal buffer states corresponding to the transition.
+    pub fn to_states(&self) -> Range<hal::buffer::State> {
+        conv::map_buffer_state(self.usage.start) ..
+        conv::map_buffer_state(self.usage.end)
+    }
+}
+
+impl Default for BufferState {
+    fn default() -> Self {
+        BufferState {
+            init: BufferUsage::empty(),
+            last: BufferUsage::empty(),
+        }
+    }
+}
+
+impl ResourceState for BufferState {
+    type Id = BufferId;
+    type Selector = ();
+    type Usage = BufferUsage;
+
+    fn query(
+        &self,
+        _selector: Self::Selector,
+    ) -> Option<Self::Usage> {
+        Some(self.last)
+    }
+
+    fn change(
+        &mut self,
+        id: Self::Id,
+        _selector: Self::Selector,
+        usage: Self::Usage,
+        output: Option<&mut Vec<PendingTransition<Self>>>,
+    ) -> Result<(), PendingTransition<Self>> {
+        let old = self.last;
+        if usage != old {
+            let pending = PendingTransition {
+                id,
+                selector: (),
+                usage: old .. usage,
+            };
+            self.last = match output {
+                Some(transitions) => {
+                    transitions.push(pending);
+                    usage
+                }
+                None =>  {
+                    if !old.is_empty() && BufferUsage::WRITE_ALL.intersects(old | usage) {
```

Comment:
I see it mentioned in the comments that only pending read-only transitions can be made here, but could we explain somewhere how writes are handled in general and how its related? i.e. I was wondering why the restriction exists (I guess one reason is that we might have mapped buffers active at this moment). Maybe (outside of this PR) we could start writing down some of these concepts in a document somewhere

Reply:
We definitely should have some piece of documentation describing this in general in `wgpu-rs`, maybe not precisely here in the deep of the tracking code :). For the most part we should refer to upstream WebGPU documentation though on the validation rules.